src/dispatch.c

```diff
@@ -82,3 +83,452 @@ natsStatus natsSub_enqueueUserMessage(natsSubscription *sub, natsMsg *msg)
+        {
+            nats_unlockDispatcher(d);
+            // Call this in case the subscription was draining.
+            natsSub_setDrainCompleteState(sub);
+
+            if (completeCB != NULL)
+                (*completeCB)(completeCBClosure);
+
+            // Subscription closed, just release
+            natsSub_release(sub);
+
+            nats_lockDispatcher(d);
+            continue;
+        }
+        else if (msg == ctrl->sub.timeout)
+        {
+            nats_unlockDispatcher(d);
+
+            // Invoke the callback with a NULL message.
+            (*messageCB)(nc, sub, NULL, messageClosure);
+
+            nats_lockDispatcher(d);
+
+            if (!draining && !closed)
+            {
+                // Reset the timedOut boolean to allow for the
+                // subscription to timeout again, and reset the
+                // timer to fire again starting from now.
+                sub->timedOut = false;
+                natsTimer_Reset(sub->timeoutTimer, sub->timeout);
+            }
+            continue;
+        }
+
+        // Fetch control messages
+        else if ((fetchStatus != NATS_OK) && !lastMessageInFetch)
+        {
+            // We drop here only if this is not already marked as last message
+            // in fetch. The last message will be delivered first.
+            natsFetchCompleteHandler fetchCompleteCB = fetch->completeCB;
+            void *fetchCompleteCBClosure = fetch->completeCBClosure;
+
+            // TODO: future: options for handling missed heartbeat, for now
+            // treat it as any other error and terminate.
+
+            nats_unlockDispatcher(d);
+            if (fetchCompleteCB != NULL)
+                (*fetchCompleteCB)(nc, sub, fetchStatus, fetchCompleteCBClosure);
+
+            // Call this blindly, it will be a no-op if the subscription
+            // was not draining.
+            natsSub_setDrainCompleteState(sub);
+            natsConn_removeSubscription(nc, sub);
+            natsMsg_Destroy(msg); // may be an actual headers-only message
+            nats_lockDispatcher(d);
+            continue;
+        }
+        else if ((fetch != NULL) && (fetchStatus == NATS_OK) && !userMsg)
+        {
+            // Fetch heartbeat. Need to set the active bit to prevent the missed
+            // heartbeat condition when the timer fires.
+            jsi->active = true;
+            natsMsg_Destroy(msg);
+            continue;
+        }
+
+        // Need to check for closed subscription again here.
+        // The subscription could have been unsubscribed from a callback
+        // but there were already pending messages. The control message
+        // is queued up. Until it is processed, we need to simply
+        // discard the message and continue.
+        else if (sub->closed)
+        {
+            natsMsg_Destroy(msg);
+            continue;
+        }
+        else if ((msg->sub == NULL) || (msg->subject == NULL) || (strcmp(msg->subject, "") == 0))
```

Comment:
I guess we could have combined with the "if" above since the action is the same, destroy message and continue.